### DISCUSSION TITLE
Is pipe channel create by `serverstart` bidirectional?

### DISCUSSION THREAD

**Discussion by linrongbin16** [neovim/neovim#24509] [Q&A]
Hi, the [serverstart](https://neovim.io/doc/user/builtin.html#serverstart) function will create a pipe (without any specified address parameters) and return the pipe address.

And I can use [sockconnect](https://neovim.io/doc/user/builtin.html#sockconnect) function to connect to it, and read/write data.

My question is, I read some tutorial about linux pipe, it looks like a linux system-level C pipe usually create a `fd[2]`, 1 is for read 1 is for write.

 I am not sure how things is going on Neovim:

1. is the pipe bidirectional? once connect successfully, both read and write is available. or there is some limitation that only server side (who create the pipe via serverstart) can write, the client side (who connect the pipe via sockconnect) can read?
2. is it available that two child process read/write data through a pipe, which is create from their parent process? or they only allow data transfer between parent and child process?

**Comment by linrongbin16** [comment]
For example, I want to write a plugin that start 2 child processes. let's say the main nvim process is P, parent. and two child processes are A and B.

In process A, I will do some calculation and write data to process B. The calculation is heavy so I want to make it async and performant.

Now I am going to follow these steps:

1. in P process, create a pipe via `serverstart`, then I got the pipe address from function returned value. let's say the address is ADDRESS.
2. write 2 lua scripts, let's say they are SA and SB (script for process A, and script for process B).
3. start 2 processes via `jobstart('nvim -l SA ADDRESS')` and `jobstart('nvim -l SB ADDRESS')`. you can see I start two lua scripts and pass them the same pipe address.
4. in SA, it will connect the pipe channel via `sockconnect('pipe', ADDRESS)`, get the channel id, let's say it's ID1. do the calculation, send data via `chansend(ID1, data)`.
5. in SB, it will also connect the pipe channel via `sockconnect('pipe', ADDRESS, {on_data=callback})`, so once there is any data received, it read and consumes them.

But I have some questions:

1. is pipe channel only available for parent-child process data transfer? or it can be used between two child processes from the same parent?

please also give any comments.

**Comment by justinmk** [comment]
> Is pipe channel create by serverstart bidirectional?

yes

> is it available that two child process read/write data through a pipe

no. that has nothing to do with Nvim, that's an OS decision.

> For example, I want to write a plugin that start 2 child processes

jobstart() creates a separate channel for each job. Each job can communicate with the Nvim server. If the jobs want to communicate with _each other_ then they need to go through the Nvim server (or set up their own orchestration outside of Nvim, which would be an overengineered waste of time).

**Reply by linrongbin16** [reply]
One more question, if I setup the channel in tcp mode, for example create via `serverstart('127.0.0.1:0')`, I believe any process on the same OS can connect to it and read/write. correct?

**Reply by justinmk** [reply]
yes, also works with named pipes ("unix domain sockets")

**Reply by linrongbin16** [reply]
thank you
